catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java

```diff
@@ -145,60 +145,68 @@ public void run() {
+        try {
+          Future<SourceResponse> future;
+          if (query.getTimeoutMillis() < 1) {
+            future = completionService.take();
+          } else {
+            future = completionService.poll(getTimeRemaining(deadline), TimeUnit.MILLISECONDS);
+            if (future == null) {
+              timeoutRemainingSources(processingDetails);
+              break;
+            }
           }
-        }
 
-        queryRequest = futures.remove(future);
-        sourceId = getSourceIdFromRequest(queryRequest);
+          queryRequest = futures.remove(future);
+          sourceId = getSourceIdFromRequest(queryRequest);
 
-        sourceResponse = future.get();
+          sourceResponse = future.get();
 
-        if (sourceResponse == null) {
-          LOGGER.debug("Source {} returned null response", sourceId);
-          executePostFederationQueryPluginsWithSourceError(
-              queryRequest, sourceId, new NullPointerException(), processingDetails);
-        } else {
-          sourceResponse = executePostFederationQueryPlugins(sourceResponse, queryRequest);
-          resultList.addAll(sourceResponse.getResults());
-          long hits = sourceResponse.getHits();
-          totalHits += hits;
-          hitsPerSource.merge(sourceId, hits, (l1, l2) -> l1 + l2);
-
-          Map<String, Serializable> properties = sourceResponse.getProperties();
-          returnProperties.putAll(properties);
-        }
-      } catch (InterruptedException e) {
-        if (queryRequest != null) {
-          // First, add interrupted processing detail for this source
-          LOGGER.debug("Search interrupted for {}", sourceId);
+          if (sourceResponse == null) {
+            LOGGER.debug("Source {} returned null response", sourceId);
+            executePostFederationQueryPluginsWithSourceError(
+                queryRequest, sourceId, new NullPointerException(), processingDetails);
+          } else if (queryRequest != null) {
+            sourceResponse = executePostFederationQueryPlugins(sourceResponse, queryRequest);
+            resultList.addAll(sourceResponse.getResults());
+            long hits = sourceResponse.getHits();
+            totalHits += hits;
+            hitsPerSource.merge(sourceId, hits, (l1, l2) -> l1 + l2);
+
+            Map<String, Serializable> properties = sourceResponse.getProperties();
+            returnProperties.putAll(properties);
+          }
+        } catch (InterruptedException e) {
+          if (queryRequest != null) {
+            // First, add interrupted processing detail for this source
+            LOGGER.debug("Search interrupted for {}", sourceId);
+            executePostFederationQueryPluginsWithSourceError(
+                queryRequest, sourceId, e, processingDetails);
+          }
+
+          // Then add the interrupted exception for the remaining sources
+          interruptRemainingSources(processingDetails, e);
+          interrupted = true;
+          break;
```

Comment:
✏️  Since we are breaking out of the loop, it turns out that we do not need to retain the interrupted state using a variable and propagate the interruption in a finally outside of the loop. We can just do the `Thread.currentThread().interrupt()` before the `beak` .. sorry, I had miss the break before.

Reply:
okay, I can put it back to what it was